File: backend/processor.js
const { PythonShell } = require("python-shell");
const path = require("path");

const ML_PATH = path.join(__dirname, "../ML"); 

// Generic function to run Python script
function runPython(mode, args = [], inputText = null) {
  return new Promise((resolve, reject) => {
    const options = {
      args: [mode, ...args],
      pythonPath: "python",
      scriptPath: ML_PATH,
    };

    const pyshell = new PythonShell("processor.py", options);
    let output = [];

    pyshell.on("message", (msg) => output.push(msg));
    pyshell.on("error", (err) => reject(err));
    pyshell.on("stderr", (err) => console.error("Python STDERR:", err));
    pyshell.on("close", () => resolve(output.join("\n")));

    if (inputText) {
      pyshell.send(inputText);
      pyshell.end(); 
    }
  });
}

async function parseFile(filePath) {
  return runPython("parse", [filePath]);
}

async function generateSummary(filePath) {
  const text = await parseFile(filePath);
  const output = await runPython("summary", [], text);
  return JSON.parse(output);
}

async function generateDiagram(input, diagramType = "flowchart", isFile = true) {
  let output;
  if (isFile) {
    output = await runPython("diagram", [input, diagramType]);
  } else {
    output = await runPython("diagram", ["-t", diagramType], input);
  }
  return JSON.parse(output);
}

module.exports = { parseFile, generateSummary, generateDiagram };

File: ML/processor.py
import sys, json
from ML_module import parse_file, generate_summary, generate_diagram

mode = sys.argv[1]

if mode in ["summary", "diagram"]:
    if len(sys.argv) > 2 and sys.argv[2] == "-t":
        text = sys.argv[3]
        diagram_type = sys.argv[4] if len(sys.argv) > 4 else "flowchart"
    elif len(sys.argv) > 2:
        file_path = sys.argv[2]
        text = parse_file(file_path)
        diagram_type = sys.argv[3] if len(sys.argv) > 3 else "flowchart"
    else:
        text = sys.stdin.read()
        diagram_type = "flowchart"
else:
    file_path = sys.argv[2]
    text = parse_file(file_path)

if mode == "parse":
    print(text)

elif mode == "summary":
    print(json.dumps(generate_summary(text), ensure_ascii=False))

elif mode == "diagram":
    print(json.dumps(generate_diagram(text, diagram_type), ensure_ascii=False))
